### DISCUSSION TITLE
I am studying the JSON and trying to configure an ESP8266 to control a water valve.

### DISCUSSION THREAD

**Discussion by Andreadf67** [RavenSystem/esp-homekit-devices#1900] [Q&A] [answered]
Thanks first for the help.
I need to control a valve connected to the GPIO5 of an esp8266 with HAA firmware.
The valve must be opened with the start button on GPIO0 and closed with the stop button on GPIO15.
I also have a switch on the GPIO2 for a full tank.
I would like to automate this:
1) when you press the start button the valve opens, if the tank is not full. If the tank is empty the valve remains open until the tank full signal GPIO2.
1) when you press the stop button the valve closes.

I tried with this Json but it doesn't work:
{"c":{"0":{"r":[{{"g":5,"v":0}]},"b":[{"g":0,"t":5}]},"a":[{"0":{"r":[{"g":5,"v":0}]},"1":{"r":[{"g":5,"v":1}]},"t":20,"b":[{"g":0,"t":1}],"f0":[{"g":15,"t":0}],"f0":[{"g":15}],"s":0,"w":0,"d":0},{"t":5,"f0":[{"g":15,"t":0}],"f1":[{"g":15}],"s":0},{"1":{"s":[{"a":1}]}}]}

**Comment by ceg117** [comment]
I think you should set the full tank switch to GPIO15, stop anyway, GPIO2 is not used in the 8266 circuit as much as possible, I attach a json, I hope it can help you.
{"c":{"o":0},"a":[{"0":{"r":[{"g":5,"v":1}]},"1":{"r":[{"g":5,"v":0}]},"t":20,"b":[],"s":0,"w":0,"d":0,"f0":[{"g":15,"t":0}],"f1":[{"g":0,"t":0}]}]}

**Reply by Andreadf67** [reply]
Thanks very much for your reply. I tried your json but it doesn't work as I thought. I will post the shema I designed as soon as possible.

**Reply by Andreadf67** [reply]
![schema](https://user-images.githubusercontent.com/98307760/197559587-65ffee85-0eab-4d94-9580-b825b45d9d6a.jpg)
![button](https://user-images.githubusercontent.com/98307760/197559614-2af4021f-7992-4f64-ba08-896419385a82.jpg)

**Comment by ceg117** [comment]
This is my idea, simplifying the operation, in the HAA firmware, with the ESP8266 circuit GPIO2 is best not used.
![1811666665906_ pic](https://user-images.githubusercontent.com/101453973/197670632-55a605bf-ac4b-4d42-b49a-f0c75f42a44d.jpg)

**Reply by Andreadf67** [reply]
Thanks very much again for your reply. I mounted your circuit and it works. Now I am studying how to add an accessory that notifies the status of the tank

**Reply by Andreadf67** [reply]
I am always tied to my initial idea and I produced this:
{"c":{"0":{"r":[{"g":5,"v":0}]}},
"a":[
{
"0":{"r":[{"g":5,"v":0}]}, RELE' VALVE
"1":{"r":[{"g":5,"v":1}]},
"t":20,
"d":0},
{ HIDDEN START
"1":{"m":[[1,0]]},
"t":1,
"h":0,
"f1":[{"g":0,"t":0}],
"f0":[{"g":0,"t":1}]
},
{ HIDDEN STOP
"1":{"m":[[1,1]]},
"t":1,
"h":0,
"f1":[{"g":15,"t":0}],
"f0":[{"g":15,"t":1}]
},
{ ACCESSORY TO CONTROL TANK LEVEL
"t":5,"ks":0
},{
"0":{"m":[[4,-10001],[4,0],[4,-10000]]},
"1":{"m":[[4,-10001],[4,1],[4,-10000],[1,0]]},
"t":5,
"h":0,
"f1":[{"g":4,"t":0}],
"f0":[{"g":4,"t":1}]
},
{"1":{"s":[{"a":1}]}}
]}

With this I solved the start and stop action and also the stop when the Tank is full.
NOW I HAVE DIFFICULTY TO UNDERSTAND HOW I CAN CHECK IF THE TANK IS FULL.
THE VALVE MUST START ONLY IF THE TANK IS EMPTY AND STOP IF THE TANK IS FULL. THANK YOU SO MUCH FOR YOUR HELP.

**Comment by ceg117** [comment]
I study it, there should be a way

**Comment by ceg117** [comment, marked as answer]
I think this should be JSON that meets your requirements, I attach the circuit diagram reference.
{"c":{"o":0},"a":[{"0":{"r":[]},"1":{"r":[{"g":5,"v":0}]},"t":5,"s":0,"f0":[{"g":14,"t":0}],"f1":[{"g":14,"t":1}]},{"0":{"r":[{"g":5,"v":1}]},"1":{"r":[]},"t":5,"s":0,"f0":[{"g":12,"t":0}],"f1":[{"g":12,"t":1}]},{"0":{"r":[{"g":5,"v":1}]},"1":{"r":[{"g":5,"v":0}]},"t":20,"b":[],"s":5,"w":0,"d":0,"f0":[{"g":12,"t":0}],"f1":[{"g":13,"t":0},{"g":14,"t":1}]}]}
![1881666965104_ pic](https://user-images.githubusercontent.com/101453973/198624263-96b655a6-f68c-40df-ace3-2866a384b9a5.jpg)

**Reply by Andreadf67** [reply]
Great and thank you very much for your help and for the time you have dedicated to me.
Best regards to you :)

**Comment by ceg117** [comment]
The MCU used in the above circuit diagram is ESP8266 12F/S, and the MCU recommends ESP826612S.

**Comment by ceg117** [comment]
Regarding some of the gpio you mentioned earlier, such as io0, io2, io15, io16 flashing HAA firmware on esp8266, these gpio have other functions, do not use it, so my json avoids these gpio.
